lib/bolt/plugin/puppetdb.rb

```diff
@@ -45,8 +45,14 @@ def resolve_reference(opts)
         targets = @puppetdb_client.query_certnames(opts['query'])
         facts = []
 
-        target_opts = opts.select { |k, _| TARGET_OPTS.include?(k) }
-        Bolt::Util.walk_vals(target_opts) do |value|
+        template = opts.delete('target_mapping') || {}
+
+        keys = Set.new(TEMPLATE_OPTS) & opts.keys
+        unless keys.empty?
+          raise Bolt::ValidationError, "PuppetDB plugin expects keys #{keys} to be set under 'target_mapping'"
```

Comment:
This is great! In addition to this, while we're here, erroring if any key besides `query` or `target_mapping` is present would also be useful. I think the main cause of the error will be if one of those options is misspelled, and especially for target_mapping running without your specified configuration could be dangerous (ie. running as root on accident)

Reply:
Updated!